### PR TITLE
Dockerfile: Fix fragmentation of some identifiers

### LINKED ISSUE #2038: The rule dockerfile.best-practice.missing-image-version.missing-image-version does not handle intermediate stages using earlier stages
**Describe the bug**
e.g.

```
FROM python:3.10.1-alpine3.15@sha256:4be65b406f7402b5c4fd5df7173d2fd7ea3fdaa74d9c43b6ebd896197a45c448 as base

# ruleid: missing-image-version
FROM base AS BUILD
```

**To Reproduce**
Use above example of docker file

**Expected behavior**
Intermediate build steps should not be affected by this rule.

**Screenshots**

**What is the priority of the bug to you?**

P2

**Environment**
semgrep docker

**Use case**
What will fixing this bug enable for you?
I currently have to run individual rules inside docker ruleset to avoid running dockerfile.best-practice.missing-image-version.missing-image-version

### LINKED ISSUE #5229: Dockerfile metavariables for image aliases are not matching
**Describe the bug**
https://semgrep.dev/playground/s/r2c-sec-research-test:missing-image-version

I am trying to update [this rule](https://semgrep.dev/editor/r/dockerfile.best-practice.missing-image-version.missing-image-version) for [this issue](https://github.com/returntocorp/semgrep-rules/issues/2038). I expected the `pattern-not-inside` clause to filter out the last test case because `$ALIAS` should match to `base`, but it does not.

**To Reproduce**
https://semgrep.dev/playground/s/r2c-sec-research-test:missing-image-version

**Expected behavior**
I expected the last test case in the example provided to not match.

**What is the priority of the bug to you?**

- [x] P2: regular bug that should get fixed

**Use case**
This will help me finish resolving this issue: https://github.com/returntocorp/semgrep-rules/issues/2038

### PR DESCRIPTION
**Pull request by mjambon**
Fixes https://github.com/returntocorp/semgrep-rules/issues/2038
Fixes https://github.com/returntocorp/semgrep/issues/5229

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: semgrep-core/src/parsing/tree_sitter/Parse_dockerfile_tree_sitter.ml [hunk 1]
```diff
@@ -307,6 +309,7 @@ let unquoted_string (env : env) (xs : CST.unquoted_string) : str =
         | `BSLASHSPACE tok -> String_content (str env tok (* "\\ " *))
         | `Imme_expa x -> expansion env x)
       xs
+    |> simplify_fragments
```
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_dockerfile_tree_sitter.ml:312]: I wonder if the type for fragments could be changed to enforce the invariant that you should call simplify_fragments.
(a bit like what Brandon did recently with his new Path.ml module).

File: semgrep-core/tests/dockerfile/from-alias-reuse.dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# MATCH:
+FROM a as b
+FROM b
+
+# MATCH:
```
Comment (aryx) [inline semgrep-core/tests/dockerfile/from-alias-reuse.dockerfile:5]: would be nice to also have negative test cases.

**Review by aryx — APPROVED** [review]

File: semgrep-core/src/parsing/tree_sitter/Parse_dockerfile_tree_sitter.ml (hunk 1 above)
Comment (mjambon) [reply semgrep-core/src/parsing/tree_sitter/Parse_dockerfile_tree_sitter.ml:312]: We could have a new type with exactly the same structure but which would have to be created with `simplify_fragments`. I think it's a bit overkill but I'll try it as an exercise.

Comment (mjambon) [reply semgrep-core/src/parsing/tree_sitter/Parse_dockerfile_tree_sitter.ml:312]: I [tried using a private type abbreviation](https://github.com/returntocorp/semgrep/pull/6144/files#r975918900) to ensure that the type `str` has no fragmented literals but it poses problems with pattern matching.

File: semgrep-core/tests/dockerfile/from-alias-reuse.dockerfile (hunk 2 above)
Comment (mjambon) [reply semgrep-core/tests/dockerfile/from-alias-reuse.dockerfile:5]: done

**Comment by mjambon** [comment]
The MacOS [job](https://github.com/returntocorp/semgrep/actions/runs/3094439350/jobs/5007814952) fails because https://opam.ocaml.org/ is down and it's needed to run to fetch package info with `opam update`. The Linux jobs don't need it because they run in a container that already has a recent version of the packages and doesn't run `opam update`.

**Comment by aryx** [comment]
Another good reason for having our own ocaml docker image!

**Review by aryx — APPROVED** [review]

**Pull request merged**